metaplex-foundation/digital-asset-rpc-infrastructure
# Add DAS integration tests

In this PR, I am adding the test suite that we use internally at Helius for testing DAS. We want to add this suite in order to make it easier to test our upstream fixes and improve the general developer experience for DAS developers. We recently revamped our test suite to make it easier and faster to run. It's currently composed of 19 tests. We have more tests internally that we will upstream as we upstream our fixes. Here are some of it's positive attributes:

1. It runs fast. The whole test suite takes a minute to run. 
2. It's easy to setup. You don't need to run the API or the ingester binaries. Just run `cargo test`. At the same, the primary functions of the ingester and API are tested in test suite. So we don't lose much test coverage through this convenience. See the README for more information. 
3. You can run individual tests by run `cargo test <test_name>`. This has made it possible to have really fast feedback loop when developing new features. 

I included setup information and more information about the tests in the README. I am going to copy this README here to make it slightly easier for reviewers to review:

# Integration Tests

This Cargo package helps us run multi-package tests in our workspace. This setup is inspired by the tokio integration test setup.

## Setup

First setup a local Postgres database and export the postgres database URL as follows:
```export DATABASE_TEST_URL=postgres://postgres@localhost/<database_name>```

Also gain access to mainnet RPCs and devnet RPCs and export the URLs as follows. Currently,
we only use these URLs for reading data and storing it locally. 

```
export DEVNET_RPC_URL=...
export MAINNET_RPC_URL=...
```

Afterwards, you can simply run the following command to run tests:
```cargo test```

## How do tests work? 

Most tests currently are configured to run as "scenario" tests. They pull test input data from mainnet/devnet
and store it locally to avoid tests breaking if mainnet/devnet data ever changes. Afterwards, they feed
the tests to the `handle_account_update` and `handle_transaction` functions of the ingester and populate
the indexed data in the database. Finally, they create an instance of the `DasApi` struct, run queries against
this struct, store the results of these queries as snapshots through the `insta` testing library and assert that
future runs of the same test produce the same snapshot. 

Note that tests do not actually run the ingester binaries and the API binaries and only test the primary internal functions
of each, e.g.  `handle_account_update` and `handle_transaction` for the ingester and functions like `search_assets` 
and `get_asset` for the binary. By following this approach, we are able to test the vast majority of the code
in a way that's easy to setup and fast to run -- avoiding to have to compile and run multiple binaries.
